source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -95,93 +101,172 @@ namespace aspect
-        for (typename std::vector<Point<dim>>::const_iterator itr = positions.begin(); itr != positions.end(); ++itr, ++index_positions)
+        const double half_h = .5;
+        for (unsigned int property_index = 0; property_index < n_particle_properties; ++property_index)
           {
-            const Tensor<1, dim, double> relative_support_point_location = (*itr - approximated_cell_midpoint) / cell_diameter;
-            for (unsigned int property_index = 0; property_index < n_particle_properties; ++property_index)
+            if (selected_properties[property_index] == true)
               {
-                double interpolated_value = c[property_index][0] +
-                                            c[property_index][1] * relative_support_point_location[0] +
-                                            c[property_index][2] * relative_support_point_location[1];
-                if (dim == 2)
+                if (use_linear_least_squares_limiter[property_index] == true)
                   {
-                    interpolated_value += c[property_index][3] * relative_support_point_location[0] * relative_support_point_location[1];
+                    c[property_index][0] = std::max(c[property_index][0], property_bounds[property_index].first);
+                    c[property_index][0] = std::min(c[property_index][0], property_bounds[property_index].second);
+
+                    double max_total_slope = std::min(c[property_index][0] - property_bounds[property_index].first,
+                                                      property_bounds[property_index].second - c[property_index][0]);
+                    double current_total_slope = 0.0;
+                    for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                      {
+                        // The slope in any one direction should not overshoot/undershoot on its own.
+                        c[property_index][i] = std::copysign(std::min(std::abs(c[property_index][i]) * 2, max_total_slope), c[property_index][i]);
+                        current_total_slope += std::abs(c[property_index][i]);
+                      }
+                    current_total_slope *= half_h;
+                    if (current_total_slope > max_total_slope)
+                      {
+                        current_total_slope = 0;
+                        for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                          {
+                            c[property_index][i] = std::copysign(std::min(std::abs(c[property_index][i]), max_total_slope / half_h), c[property_index][i]);
+                            current_total_slope += std::abs(c[property_index][i]);
+                          }
+                        current_total_slope *= half_h;
+                        if (current_total_slope > max_total_slope)
+                          {
+                            double change_in_slope = (current_total_slope - max_total_slope)/(dim * half_h);
```

Comment:
My question here is: Why do you add a constant amount of slope to all components? This could push some of the components to having a different sign, which creates the problems you try to solve below  (lines 208 ff.). Would it not make more sense to compute the ratio between current_total_slope and max_total_slope, and then divide each `c` component by this ratio? This should also guarantee that you end up with at most `max_total_slope`, but it will apply the majority of the change to the largest components of c (i.e. the components that cause the overshoot in the first place), and it can never change the sign of the components.

I think the case that makes me uncomfortable with your current algorithm is the following:
Assume the properties you use as an example below ([.5, 1, .04, .01]). The large values of c[1] and c[2] cause the overshoot. However, your correction will be equally applied to c[1], c[2], *and* c[3]. In other words the linear plane is mostly tilted in x (most) and y (a little) direction, but after the correction the plane is also significantly tilted in z-direction. This seems to me like a bad property of the algorithm. If instead the amount of correction would scale with the size of c then we would not see properties developing large gradients in directions they had no gradient before.